### PR TITLE
feat: Implement create, delete, clone, regen, load and unload

### PR DESCRIPTION
**Pull request by benwoo1110**
Implemented the basic world generation commands.

- `/mv clone <world> <new-world-name>`
- `/mv confirm`
- `/mv create <name> <env> --seed [seed] --generator [generator[:id]] --world-type [worldtype] --adjust-spawn --no-structures`
- `/mv delete <world>`
- `/mv load <world>`
- `/mv unload <world>`
- `/mv regen <world> --seed [seed] --keep-gamerules`

Test Build:
[Multiverse-Core-5.0.0-SNAPSHOT.zip](https://github.com/Multiverse/Multiverse-Core/files/10687166/Multiverse-Core-5.0.0-SNAPSHOT.zip)

### CODE REVIEW

File: src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    @Subcommand("clone")
+    @CommandPermission("multiverse.core.clone")
+    @Syntax("<world> <new-world-name>")
+    @CommandCompletion("@mvworlds:type=includeunloaded @empty")
+    @Description("Clones a world.")
+    public void onCloneCommand(CommandIssuer issuer,
+
+                               @NotNull
+                               @Syntax("<world>")
+                               @Description("Current multiverse world.")
```
Comment (zax71) [inline src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java:32]: World to clone could be more verbose? I read that as the world I am currently in

File: src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                               @Description("New cloned world name.")
+                               String newWorldName
+    ) {
+        issuer.sendMessage(String.format("Cloning world '%s' to '%s'...", world.getName(), newWorldName));
+
+        if (!this.plugin.getMVWorldManager().cloneWorld(world.getName(), newWorldName)) {
+            issuer.sendMessage(String.format("%sWorld could not be cloned! See console for more details.", ChatColor.RED));
+            return;
+        }
+        issuer.sendMessage(String.format("%sWorld cloned!", ChatColor.GREEN));
```
Comment (zax71) [inline src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java:47]: Again, "World successfully cloned" would be more verbose

File: src/main/java/com/onarandombox/MultiverseCore/commands/DeleteCommand.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    @CommandPermission("multiverse.core.delete")
+    @Syntax("<world>")
+    @CommandCompletion("@mvworlds:type=includeunloaded")
+    @Description("Deletes a world on your server PERMANENTLY.")
+    public void onDeleteCommand(BukkitCommandIssuer issuer,
+
+                                @Single
+                                @Conditions("worldname:type=includeunloaded")
+                                @Syntax("<world>")
+                                @Description("Multiverse world you want to delete.")
```
Comment (zax71) [inline src/main/java/com/onarandombox/MultiverseCore/commands/DeleteCommand.java:33]: Don't think being a Multiverse world is required to be defined here

File: src/main/java/com/onarandombox/MultiverseCore/commands/UnloadCommand.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Subcommand("unload")
+    @CommandPermission("multiverse.core.unload")
+    @CommandCompletion("@mvworlds")
+    @Syntax("<world>")
+    @Description("Unloads a world from Multiverse. This does NOT remove the world folder. This does NOT remove it from the config file.")
+    public void onUnloadCommand(BukkitCommandIssuer issuer,
+
+                                @Syntax("<world>")
+                                @Description("Name of world you want to unload.")
```
Comment (zax71) [inline src/main/java/com/onarandombox/MultiverseCore/commands/UnloadCommand.java:28]: Name of **the** world you want to unload.

File: src/main/java/com/onarandombox/MultiverseCore/commandtools/MVCommandCompletions.java [hunk 5]
```diff
@@ -44,9 +48,24 @@ private Collection<String> suggestMVWorlds(BukkitCommandCompletionContext contex
...
-        return this.plugin.getMVWorldManager().getMVWorlds()
-                .stream()
-                .map(MVWorld::getName)
-                .collect(Collectors.toList());
+        String type = context.getConfig("type", "loaded");
+
+        List<String> worlds = new ArrayList<>();
+
+        switch (type) {
+            case "includeunloaded":
```
Comment (zax71) [inline src/main/java/com/onarandombox/MultiverseCore/commandtools/MVCommandCompletions.java:56]: Could be batter to make this camelCase

**Review by zax71 — CHANGES_REQUESTED** [review]
Mostly just grammatical errors. To my untrained eye, the code looks fine :)

**Comment by NeverGrowUp** [comment]
Tested snapshot. Had a few limitations to testing as i couldnt change mv world gamerules to be able to test --keep-gamerules. Comments at current point:

- Doing `/mv regen test1 --seed 1234` (prev seed was 1111) resulted in a random seed.
- Couldnt test `--keep-gamerules` as `/mv gamerules` isnt available.
- Had a single line error after doing `/mv debug 3`, `[Multiverse-Core] [ACF] Could not find command condition debuglevel for onChangeDebugCommand::arg1` But this only occured once and could not reproduce again.

All other commands/actions went well, including `pl:<name>`

File: src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java (hunk 1 above)
Comment (dumptruckman) [reply src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java:32]: "World to clone".. yeah might read better.

File: src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java (hunk 2 above)
Comment (dumptruckman) [reply src/main/java/com/onarandombox/MultiverseCore/commands/CloneCommand.java:47]: Not sure about this one. Verbosity isn't necessarily the goal. Really, I'd prefer the most concise message that conveyed the meaning in the most helpful way.

File: src/main/java/com/onarandombox/MultiverseCore/commands/DeleteCommand.java (hunk 3 above)
Comment (dumptruckman) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DeleteCommand.java:33]: Yeah take it or leave it I guess. Thought I do think this command only works on worlds managed by Multiverse?

File: src/main/java/com/onarandombox/MultiverseCore/commandtools/MVCommandConditions.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        switch (type) {
+            // Worlds that are loaded
+            case "loaded":
+                if (!this.plugin.getMVWorldManager().isMVWorld(worldName)) {
+                    throw new ConditionFailedException("World with name '" + worldName + "' does not exist or is not loaded!");
+                }
+                break;
+            // World that are loaded or unloaded
+            case "includeunloaded":
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseCore/commandtools/MVCommandConditions.java:43]: How about `"both"`?

Comment (benwoo1110) [reply src/main/java/com/onarandombox/MultiverseCore/commandtools/MVCommandConditions.java:43]: hmm good idea

File: src/main/java/com/onarandombox/MultiverseCore/commands/DeleteCommand.java (hunk 3 above)
Comment (benwoo1110) [reply src/main/java/com/onarandombox/MultiverseCore/commands/DeleteCommand.java:33]: essentially if you are using an mv command its has to be an mv world (apart from import), it only works on mv worlds, so ig no need to specify **again** that it needs to be an mv world.

**Comment by benwoo1110** [comment]
okay fixed all the concerns, gamerules commands will be implemented later, heard @zax71 is working on it :eyes:

**Comment by zax71** [comment]
> okay fixed all the concerns, gamerules commands will be implemented later, heard @zax71 is working on it eyes

Yep, correct! I will try to finalise it for PR today. If not, then I cannot work on it until Monday.

**Pull request merged**